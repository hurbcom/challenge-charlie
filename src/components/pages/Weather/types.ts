export interface CurrentWeatherProps {
  wind: {
    speed: number
    deg: number
  }
  weather: [
    {
      main: string
      description: string
      icon: number
    }
  ]
  main: {
    temp: number
    pressure: number
    humidity: number
  }
}

export interface ForecastWeatherProps {
  list: {
    main: {
      temp: number
    }
  }[]
}

export interface LocationProps {
  results: {
    components: {
      city: string
      state?: string
      country: string

    }
    geometry: {
      lat: number
      lng: number
    }
  }[]
}