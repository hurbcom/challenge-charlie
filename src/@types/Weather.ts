export type WeatherData = {
  coord: {
    lon: number
    lat: number
  }
  weather: [
    {
      id: number
      main: string
      description: string
      icon: string
    },
  ]
  base: string
  main: {
    temp: number
    feels_like: number
    temp_min: number
    temp_max: number
    pressure: number
    humidity: number
    sea_level: number
    grnd_level: number
  }
  visibility: number
  wind: {
    speed: number
    deg: number
    gust: number
  }
  clouds: {
    all: number
  }
  dt: number
  sys: {
    country: string
    sunrise: number
    sunset: number
  }
  timezone: number
  id: number
  name: string
  cod: number
}

export interface Today {
  main: {
    tempC: string
    tempF: string
    pressure: number
    humidity: number
  }

  wind: {
    speed: number
    deg: number
  }

  weather: {
    icon: string
    description: string
  }
}
