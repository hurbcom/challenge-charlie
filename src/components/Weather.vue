<template>
    <div class="wrapper-weather-container">
        <div
            class="d-flex weather-row"
            :class="[
                {'today': today},
                backgroundColorProps
            ]"
        >
            <div class="icon icon-weather">
                <span
                    class="icon"
                    :class="{ 'hide': !today }"
                    :data-icon="generateIcon(weather.description)"
                >
                </span>
            </div>

            <div class="details-weather d-flex flex-column">
                <p class="date">{{ title }}</p>
                <p
                    class="temperature"
                    v-if="showCelsius"
                    @click="toggleTemperature('F')"
                >
                    {{ weather.tempC }}ºC
                </p>
                <p
                    class="temperature"
                    v-else
                    @click="toggleTemperature('C')"
                >
                    {{ weather.tempF }}ºF
                </p>

                <div :class="{ 'hide': !today }">
                    <p class="state">{{ weather.description }}</p>

                    <p class="wind">Vento: {{ weather.wind }}</p>
                    <p class="humidty">Humidade: {{ weather.humidity }}</p>
                    <p class="pressure">Pressão: {{ weather.pressure }}</p>
                </div>
            </div>
        </div>
    </div>
</template>

<script>
import temperature from './../constants/temperature';
import temperatureIcons from './../constants/temperatureIcons';

export default {
    name: 'Weather',
    props: {
        weather: {
            type: Object,
            required: true,
            default: () => ({
                tempC: '',
                tempF: '',
                description: '',
                pressure: '',
                humidity: '',
                wind: '',
            }),
        },
        title: {
            type: String,
            required: true,
            default: '',
        },
        showCelsius: {
            type: Boolean,
            required: false,
            default: true
        }
    },
    computed: {
        today() {
            return this.title === 'Hoje';
        },
        backgroundColorProps() {
            let color = 'gray';

            if (!this.weather.tempC) {
                return this.generateColorTones(color);
            }

            if (this.weather.tempC <= temperature.LOWER) {
                color = 'blue';
            } else  if (this.weather.tempC > temperature.LOWER && this.weather.tempC < temperature.HIGHER) {
                color = 'yellow';
            } else if (this.weather.tempC >= temperature.HIGHER) {
                color = 'red';
            }

            return this.generateColorTones(color);
        },
    },
    methods: {
        generateColorTones(color) {
            const tones = [1, 2 , 3, 4];
            const tone = tones[Math.floor(Math.random() * tones.length)];
            return `${color} tone-${tone}`;
        },
        toggleTemperature(type) {
            this.$emit('changeTemperatureType', type);
        },
        generateIcon(temperature) {
            let icon = ')';

            Object.keys(temperatureIcons).forEach(temperatureLabel => {
                if (temperature.includes(temperatureLabel)) {
                    icon = temperatureIcons[temperatureLabel];
                }
            })

            return icon;
        },
    },
}
</script>

<style lang="scss" scoped>
    .wrapper-weather-container {
        .hide {
            display: none;
        }

        .weather-row {
            padding: 10px 0 20px 0;
            height: 20vh;

            &.today {
                height: 45vh;
            }

            &.gray {
                &.tone-1 { background-color: rgba(211, 211, 211, 0.9); }
                &.tone-2 { background-color: rgba(201, 201, 201, 0.9); }
                &.tone-3 { background-color: rgba(148, 148, 148, 0.9); }
                &.tone-4 { background-color: rgba(124, 124, 124, 0.9); }
            }

            &.yellow {
                &.tone-1 { background-color: rgba(255, 255, 0, .9); }
                &.tone-2 { background-color: rgba(216, 216, 0, .9); }
                &.tone-3 { background-color: rgba(178, 178, 0, .9); }
                &.tone-4 { background-color: rgba(140, 140, 0, .9); }
            }

            &.red {
                &.tone-1 { background-color: rgba(255, 0, 0, .9); }
                &.tone-2 { background-color: rgba(216, 0, 0, .9); }
                &.tone-3 { background-color: rgba(178, 0, 0, .9); }
                &.tone-4 { background-color: rgba(140, 0, 0, .9); }
            }

            &.blue {
                &.tone-1 { background-color: rgba(0, 192, 255, .9); }
                &.tone-2 { background-color: rgba(0, 172, 255, .9); }
                &.tone-3 { background-color: rgba(0, 113, 216, .9); }
                &.tone-4 { background-color: rgba(0, 73, 140, .9); }
            }
        }

        .icon-weather {
            width: 60%;
            text-align: center;

            .icon::before {
                font-size: 15rem;
                color: #fff;
            }
        }

        .details-weather {
            p {
                color: #fff;
                font-size: 1.3rem;
                font-weight: 400;

                &.temperature {
                    cursor: pointer;
                }

                &.date {
                    text-transform: uppercase;
                }

                &.state {
                    padding: 15px 0 20px;
                    text-transform: capitalize;
                }

                &.wind,
                &.humidty,
                &.pressure {
                    font-size: 1rem;
                }
            }
        }
    }
</style>