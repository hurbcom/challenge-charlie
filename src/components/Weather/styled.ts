import styled from "styled-components";

export const WeatherContainer = styled.div`
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;

    width: 60%;
    height: 100%;
    max-height: 100vh;
`;

export const WeatherHeader = styled.header`
    display: flex;
    align-items: center;
    justify-content: center;

    width: 100%;
    height: 16%;

    background-color: var(--gray);

    padding: 0 20px;
`;

export const CityInput = styled.input`
    width: 100%;
    height: 100%;

    background-color: var(--gray);
    border: none;

    font-size: 2rem;
    padding: 0 0.5rem;
`;

export const WeatherTodayWrapper = styled.div<{ BgColor: string }>`
    width: 100%;
    height: 52%;
    background-color: ${(props) => props.BgColor};
`;

export const WeatherTodayInnerWrapper = styled.div`
    display: flex;
    align-items: center;
    justify-content: space-around;

    width: 100%;
    height: 100%;
`;

export const WeatherTodayIcon = styled.img``;

export const CurrentWeatherWrapper = styled.div`
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    justify-content: center;
`;

export const CurrentWeatherTempWrapper = styled.div`
    display: flex;
    flex-direction: column;

    font-size: 1.5rem;
    font-weight: bold;
`;

export const CurrentWeatherDescription = styled.span`
    font-size: 1.5rem;

    margin-bottom: 1rem;
`;

export const CurrentWeatherInfos = styled.span`
    font-size: 1.2rem;
`;

export const ForecastWrapper = styled.div<{ BgColor: string }>`
    display: flex;
    justify-content: space-around;
    align-items: flex-end;
    flex-direction: column;

    font-size: 1.5rem;
    font-weight: bold;

    padding: 0 1rem;

    width: 100%;
    height: 16%;
    background-color: ${(props) => props.BgColor};
`;

// export const WeatherTomorrowWrapper = styled.div<{ BgColor: string }>`
//     width: 100%;
//     height: 16%;
//     background-color: ${(props) => props.BgColor};
// `;

// export const WeatherDayAfterTomorrowWrapper = styled.div<{ BgColor: string }>`
//     width: 100%;
//     height: 16%;
//     background-color: ${(props) => props.BgColor};
// `;

export const TemperatureWrapper = styled.div`
    display: flex;
    align-items: center;
`;
