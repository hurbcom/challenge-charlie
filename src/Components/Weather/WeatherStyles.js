import styled from "styled-components";

export const StyledWeather = styled.div`
    width: 100%;
    box-shadow: 1px 1px 2px 2px #0000002b;
    color: #fff;
    padding: 25px 45px 30px;
    display: flex;
    justify-content: flex-end;
    align-items: center;
    flex-direction: column;
    background: linear-gradient(
        to right,
        rgba(40, 40, 40, 0.9) 0%,
        rgba(70, 70, 70, 0.9) 100%
    );
    @media (min-width: 1024px) {
        position: relative;
        flex-direction: row;
        padding: 25px 85px 60px;
    }
    .icon {
        font-size: 180px;
        @media (min-width: 1024px) {
            position: absolute;
            top: 50%;
            transform: translateY(-50%);
            left: 20px;
        }
        @media (min-width: 1024px) {
            font-size: 250px;
            left: 50px;
        }
        @media (min-width: 1366px) {
            font-size: 300px;
            left: 70px;
        }
    }
    .info {
        width: 100%;
        @media (min-width: 1024px) {
            width: 345px;
        }
        .details {
            margin-top: 30px;
            @media (min-width: 1024px) {
                margin-top: 55px;
            }
        }
        h2 {
            text-transform: uppercase;
        }
        h2,
        h3 {
            margin-bottom: 15px;
            line-height: 100%;
            font-size: 27px;
            @media (min-width: 1024px) {
                font-size: 32px;
            }
        }
        h3 {
            text-transform: capitalize;
        }
        p {
            @media (min-width: 1024px) {
                font-size: 22px;
            }
        }
    }
    .details,
    .icon {
        display: none;
    }
    &:hover {
        padding-bottom: 70px;
        @media (min-width: 1366px) {
            padding-bottom: 125px;
        }
        .details,
        .icon {
            display: block;
        }
    }
`;
