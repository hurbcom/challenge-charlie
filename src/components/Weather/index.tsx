import { FormEvent, useCallback, useEffect, useState } from "react";
import styled, { css } from "styled-components";

import AppHeader from "components/AppHeader";
import WeatherList from "components/WeatherList";
import useDailyForecast from "hooks/useDailyForecast";
import { AppCoordinatesType } from "hooks/useGeocode";

const WeatherApp = styled.div`
    display: flex;
    flex-direction: column;
    align-items: center;
    height: fit-content;
    width: 100%;

    & > * {
        max-width: 340px;
        width: 100%;
    }
`;

const Error = styled.div`
    background-color: red;
    color: white;
    position: absolute;
    bottom: 0;
`;

export const Weather = () => {
    const [search, setSearch] = useState<AppCoordinatesType>(
        {} as AppCoordinatesType
    );

    const { forecast, error, loadForecast } = useDailyForecast();

    const handleChange = (coords: AppCoordinatesType) => {
        setSearch(coords);
    };

    useEffect(() => {
        const handler = setTimeout(() => {
            if (search.latitude && search.longitude) loadForecast(search);
        }, 500);

        return () => {
            clearTimeout(handler);
        };
    }, [search, search]);

    return (
        <WeatherApp>
            <AppHeader onChange={handleChange} />
            {forecast ? <WeatherList data={forecast} /> : null}
            {error ? <Error>{error}</Error> : null}
        </WeatherApp>
    );
};

export default Weather;
