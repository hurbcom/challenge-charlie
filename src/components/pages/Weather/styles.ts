import  styled  from 'styled-components'

export const Container = styled.div<{ backgroundUrl: string }>`
  width: 100vw;
  height: 100vh;
  display: flex;
  flex-direction: column;
  ${({ backgroundUrl }) => {
       return `background-image: url(${backgroundUrl});`
    }}
  align-items: center;
`

export const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  width: 70vw;
  margin-top: 10px;
`