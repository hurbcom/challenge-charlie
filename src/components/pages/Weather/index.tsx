import React, { useEffect, useState } from "react";
import { getBingBackgroung } from "../../../services/getBingBackground";
import { getLocationByCoordinates, getLocationByLocalName } from "../../../services/getLocation";
import { getCurrentWeather, getForecastForNextDays } from "../../../services/getWeather";
import { Colors } from "../../../utils/colors";
import Banner from "../../atoms/Banner";
import Input from "../../atoms/Input";
import Icon from  '../../../assets/icons/sun.svg'

import { Container, Wrapper } from "./styles";

const Weather: React.FC = () => {
  const [locationName, setLocationName] = useState('')
  const [currentWeather, setCurrentWeather] = useState('')
  const [forecast, setForecast] = useState('')
  const [bingBackground, setBingBackground] = useState('')
  const [localName, setLocalName] = useState('')

  const callLocationByCoordinates = async (latitude: number, longitude: number) =>  {
    const response = await getLocationByCoordinates(latitude, longitude)
    setLocationName(response)
  }

  const callCurrentWeather = async (latitude: number, longitude: number) =>  {
    const response = await getCurrentWeather(latitude, longitude)
    setCurrentWeather(response)
  }

  const callForecastNextDays = async (latitude: number, longitude: number) =>  {
    const response = await getForecastForNextDays(latitude, longitude)
    setForecast(response)
  }

  const getBackground = async () => {
    const response = await getBingBackgroung()
    setBingBackground(response[0].url)
  }

  const callLocationByName = async () =>  {
    if (localName !== "") {
     const response = await getLocationByLocalName(localName)
     setLocationName(response)
    }
  }

  useEffect( () => {
    navigator.geolocation.getCurrentPosition( position => {
      callLocationByCoordinates(position.coords.latitude, position.coords.longitude)
      callCurrentWeather(position.coords.latitude, position.coords.longitude)
      callForecastNextDays(position.coords.latitude, position.coords.longitude)
    })
    getBackground()
  }, [] )

  return (
  <Container backgroundUrl={bingBackground}>
    <Wrapper>
      <Input value={localName} onChange={(value) => setLocalName(value)} onBlur={callLocationByName}/>
      <Banner
        imgSrc={Icon}
        dayTitle="HOJE"
        temperature="23"
        temperatureDescription="Calor"
        additionalInfo={{
          vento: 'NO 6.4km/h',
          umidade: '78%',
          pressao: '100 3h PA'
        }}
        backgroundColor={Colors.MEDIUM_YELLOW}/>
      <Banner dayTitle="AMANHÃ"  temperature="25" backgroundColor={Colors.DARK_YELLOW}/>
      <Banner dayTitle="DEPOIS DE AMANHÃ"  temperature="26" backgroundColor={Colors.DARKEST_YELLO}/>
    </Wrapper>

  </Container>
  )
};

export default Weather
