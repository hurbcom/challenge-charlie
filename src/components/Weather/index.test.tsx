import { render } from "@testing-library/react";
import Weather from ".";

jest.useFakeTimers();

jest.mock("../../hooks/useGeocode", () => {
    return jest
        .fn()
        .mockImplementation(() => ({
            result: {},
            translate: jest.fn(),
        }))
        .mockImplementationOnce(() => ({
            result: {
                latitude: -27.1,
            },
            translate: jest.fn().mockImplementation(() =>
                Promise.resolve({
                    latitude: 27.5,
                    longitude: 48.5,
                    city: "Florianópolis",
                    state: "Santa Catarina",
                })
            ),
        }))
        .mockImplementationOnce(() => ({
            result: {
                latitude: -27.5,
                longitude: -48.5,
            },
            translate: jest.fn().mockImplementation(() =>
                Promise.resolve({
                    latitude: -27.5,
                    longitude: -48.5,
                    city: "Florianópolis",
                    state: "Santa Catarina",
                })
            ),
        }));
});

jest.mock("../../hooks/useDailyForecast", () => {
    return jest.fn(() => ({
        forecast: [],
        loadForecast: jest
            .fn()
            .mockImplementation(() => Promise.resolve({ daily: [] })),
    }));
});

describe("<Weather />", () => {
    test("renders the weather app without data", () => {
        const { asFragment } = render(<Weather />);
        expect(asFragment).toMatchSnapshot();
        jest.runAllTimers();
    });
});
