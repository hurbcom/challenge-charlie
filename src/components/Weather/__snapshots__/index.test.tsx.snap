// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Weather /> renders the weather app without data 1`] = `[Function]`;
